### PR TITLE
Get filetree picker url parameters via own method

### PR DESCRIPTION
**Pull request by rabauss**
Get the picker url parameters via own method, to get the possibility to overwrite/extend it in own extensions.

### CODE REVIEW

**Comment by aschempp** [comment]
Can you explain why you want to override the `FileTree`? And can you not just set these properties on the object?

**Comment by rabauss** [comment]
We want to transfer more context data (table, field, id, providers) to the picker url, so that we can use it for calls to our external asset managment system. The extension is still in progress - but kind of similar to https://github.com/terminal42/contao-bynder/

**Comment by Toflar** [comment]
Maybe you should have your completely own picker instead of extending the core one then?😊

**Comment by aschempp** [comment]
I guess https://github.com/contao/core-bundle/pull/1171 is what you need. Until then you should create your own widget and not extend the core widget.

**Comment by rabauss** [comment]
Well, yes we could use our own picker but we wanted to replace the filepicker with custom asset picker from external system - and so we thought this could be achived best by replacing/extending the core filetree without too much duplicate code for similar widget.

**Comment by aschempp** [comment]
You should probably not replace the file widget, but add a new picker to the popup? We did something similar in https://github.com/terminal42/contao-bynder

**Comment by rabauss** [comment]
I know your extension. But we want to support the core filemanager and external DAm just for the first insert. After picking a file from the external DAM we want to disable the core filepicker. I think this is only possible via overwriting the core filetree, isn't it?

**Comment by aschempp** [comment]
There are picker configuration options to limit the available tabs in the back end. Not sure if you can specify them for the filePicker though.

**Comment by rabauss** [comment]
At the moment we set `$extras['providers'] = 'ourPicker'` in the fileTreeWidget if the FileModel has an ID of the external DAM - therefor among other variables for the pickerUrl (like fieldname, table), we added this PR!
Of course we could write an own widget (like we did at the moment), but the only difference to the fileTree would be the extra picker url variables.

**Comment by aschempp** [comment]
I understand now. In the regular `extra => dcawizard => []` field you can add unlimited props for the pickers. Maybe we should introduce something similar for this case?

Or rather have a universal picker and use it everywhere 😇

**Comment by rabauss** [comment]
Yes, an universal picker in the core would be the best :-)
We have an internal package adapted to the https://github.com/contao/core-bundle/pull/1171 but there is too much legacy code inside to reuse.

**Comment by leofeyer** [comment]
@rabauss @aschempp What is the status of this PR now that we have added a universal picker in #714?

**Comment by rabauss** [comment]
The main problem remains in the widget and also in the new picker widget.
https://github.com/contao/contao/blob/2aeda7f9c41eb8725c0c9b47e3ed3303c9daf7a8/core-bundle/src/Resources/contao/widgets/Picker.php#L150

The url parameters (`$extras`) for the picker cannot be overwritten/extended, if someone `extends` the class! I think I should rebase this PR and I will open another one for `widgets/Picker.php`

File: core-bundle/src/Resources/contao/widgets/FileTree.php [hunk 1]
```diff
@@ -420,6 +400,41 @@ public function generate()
 		return $return;
 	}
 
+	/**
+	 * Return the extra parameters for picker url
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/widgets/FileTree.php]:
```suggestion
	 * Return the extra parameters for the picker URL
```

**Review by leofeyer — CHANGES_REQUESTED** [review]

**Review by aschempp — APPROVED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @rabauss.

**Pull request merged**